File: sdks/csharp/Sauron.Tests/WireFixtureIo.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.CompilerServices;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Sauron.Tests;

/// <summary>
/// Writer for <c>sdks/wire-fixtures/&lt;sdk&gt;.json</c> — the envelopes the backend's
/// <c>cargo test -p sauron-core --test sdk_wire_conformance</c> feeds through the REAL
/// <c>serde</c> deserializer.
/// </summary>
/// <remarks>
/// <para>Two categories are pinned so that regenerating is a NO-OP:</para>
/// <list type="number">
///   <item>the intrinsically dynamic fields (<c>timestamp</c>, <c>event_id</c>, …);</item>
///   <item>everything the <b>toolchain</b> supplies rather than the SDK — stack-frame
///   identity strings (the test assembly's own names) and the host/runtime values in
///   <c>context.os</c> / <c>.runtime</c> / <c>.device</c>. Without this a .NET SDK
///   upgrade rewrote a committed file with no wire change at all, which makes a CI
///   diff gate noisy and leaves a tracked file dirty after a plain test run.</item>
/// </list>
/// <para>
/// What is deliberately NOT normalized is the part that proves something: item shape,
/// key set, nullability, and the frame COUNT.
/// </para>
/// </remarks>
internal static class WireFixtureIo
{
    private const string Timestamp = "2026-07-12T10:30:00.123Z";

    private static readonly Dictionary<string, string> StringSubs = new()
    {
        ["timestamp"] = Timestamp,
        ["sent_at"] = Timestamp,
        ["event_id"] = "0123456789abcdef0123456789abcdef",
        ["session_id"] = "sess_fixture",
        ["device_id"] = "3f2504e0-4f89-41d3-9a0c-0305e82c3301",
        ["workflow_id"] = "wf_fixture",
        ["raw_stacktrace"] = "<normalized>",
        ["build_id"] = "<normalized>",
        ["isolate_dso_base"] = "<normalized>",
    };

    /// <summary>Stack-frame identity: where the test ran, not what the SDK emits.</summary>
    private static readonly Dictionary<string, string> FrameIdentity = new()
    {
        ["function"] = "<fn>",
        ["module"] = "<module>",
        ["filename"] = "<file>",
        ["abs_path"] = "<file>",
    };

    /// <summary>
    /// <c>"&lt;parent&gt;.&lt;key&gt;"</c> paths carrying host- or runtime-derived values.
    /// <c>context.device</c> / <c>.os</c> / <c>.runtime</c> are free-form
    /// <c>serde_json::Value</c> on the wire, so their contents prove nothing — while
    /// <c>runtime.version</c> is the .NET version. <c>runtime.name</c> is left alone
    /// deliberately: it is an SDK constant, not a host value.
    /// </summary>
    private static readonly HashSet<string> HostDerived = new()
    {
        "os.name",
        "os.version",
        "runtime.version",
        "device.family",
        "device.model",
        "device.arch",
    };

    /// <summary>Write one captured envelope body as this SDK's committed wire fixture.</summary>
    public static string Write(string sdk, string envelopeJson, [CallerFilePath] string sourceFile = "")
    {
        var node = JsonNode.Parse(envelopeJson) ?? throw new InvalidOperationException("empty envelope body");
        Normalize(node, string.Empty, string.Empty);

        // Sauron.Tests/ -> csharp/ -> sdks/. Derived from the SOURCE path, not the test
        // assembly's bin/ directory, so it does not depend on the build layout.
        var testsDir = Path.GetDirectoryName(sourceFile)!;
        var sdksDir = Path.GetFullPath(Path.Combine(testsDir, "..", ".."));
        var path = Path.Combine(sdksDir, "wire-fixtures", $"{sdk}.json");
        Directory.CreateDirectory(Path.GetDirectoryName(path)!);
        File.WriteAllText(path, node.ToJsonString(new JsonSerializerOptions { WriteIndented = true }) + "\n");
        return path;
    }

    private static void Normalize(JsonNode node, string key, string parentKey)
    {
        switch (node)
        {
            case JsonObject obj:
                foreach (var name in new List<string>(obj.Select(p => p.Key)))
                {
                    var child = obj[name];
                    if (child is null)
                        continue;
                    var replacement = Replacement(name, key, child);
                    if (replacement is not null)
                        obj[name] = replacement;
                    else
                        Normalize(child, name, key);
                }
                break;
            case JsonArray arr:
                // Array children keep the container's key AND its parent, so a frame
                // inside `stacktrace: [...]` is still seen as living under it.
                for (var i = 0; i < arr.Count; i++)
                {
                    var child = arr[i];
                    if (child is null)
                        continue;
                    var replacement = Replacement(key, parentKey, child);
                    if (replacement is not null)
                        arr[i] = replacement;
                    else
                        Normalize(child, key, parentKey);
                }
                break;
        }
    }

    /// <summary>The pinned value for a leaf, or null when the leaf is kept verbatim.</summary>
    private static JsonNode? Replacement(string key, string parentKey, JsonNode node)
    {
        if (node is not JsonValue value)
            return null;
        if (value.TryGetValue<string>(out _))
        {
            if (HostDerived.Contains($"{parentKey}.{key}"))
                return JsonValue.Create("<host>");
            if (FrameIdentity.TryGetValue(key, out var frame))
                return JsonValue.Create(frame);
            if (StringSubs.TryGetValue(key, out var sub))
                return JsonValue.Create(sub);
            return null;
        }
        if (value.TryGetValue<int>(out _))
        {
            if (key == "lineno")
                return JsonValue.Create(42);
            if (key == "colno")
                return JsonValue.Create(13);
        }
        // JSON `null` is left alone: nullability is part of what the fixture proves.
        return null;
    }
}

File: sdks/js/test/wire-fixture-io.ts
import { mkdirSync, writeFileSync } from 'node:fs';
import { dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

/**
 * Shared writer for `sdks/wire-fixtures/<sdk>.json` — the envelopes the backend's
 * `sdk_wire_conformance` test feeds through the real `serde` deserializer.
 *
 * ## What is pinned, and why
 *
 * A fixture is only useful if regenerating it is a NO-OP. Two things follow from
 * that, and the second one is the whole reason this file grew:
 *
 * 1. The intrinsically dynamic fields (`timestamp`, `event_id`, …) are pinned.
 * 2. So is everything **the toolchain supplies rather than the SDK**. The
 *    fixture is written from inside vitest running on Node, so without this the
 *    file recorded vitest's internal frame names (`chunk-hooks.js`, `runSuite`,
 *    `processTicksAndRejections`), the host kernel version and the Node version.
 *    Bumping vitest or Node then rewrote a committed file **with no wire change
 *    at all** — which makes a CI diff gate permanently noisy and trains
 *    reviewers to wave fixture diffs through, and means the suites dirty a
 *    tracked file on machines that only ran the tests.
 *
 * What is deliberately NOT normalized is the part that actually proves
 * something: item shape, key set, nullability, and the frame COUNT. A frame's
 * identity strings say nothing about whether the envelope deserializes; that
 * there are 9 frames, each with these keys and these nulls, does.
 *
 * Residual: frame count and null-vs-string per frame still come from the test
 * runner, so a vitest/Node MAJOR change can legitimately move them. That is
 * deterministic per lockfile — CI and a developer both run `npm ci` — so it
 * shows up only alongside a dependency bump, where a fixture diff is expected
 * and reviewable. See `sdks/wire-fixtures/README.md`.
 */

const FIXED = {
  timestamp: '2026-07-12T10:30:00.123Z',
  eventId: '0123456789abcdef0123456789abcdef',
  sessionId: 'sess_fixture',
  deviceId: '3f2504e0-4f89-41d3-9a0c-0305e82c3301',
  workflowId: 'wf_fixture',
  lineno: 42,
  colno: 13,
} as const;

/**
 * Stack-frame identity. These come from wherever the test happened to run, not
 * from the SDK: under vitest they are the runner's own internals.
 */
const FRAME_IDENTITY: Record<string, string> = {
  function: '<fn>',
  module: '<module>',
  filename: '<file>',
  abs_path: '<file>',
};

/**
 * `<parent>.<key>` paths whose value is host- or toolchain-derived.
 *
 * `context.device` / `.os` / `.runtime` are free-form `serde_json::Value` on the
 * wire (`envelope.rs`: `pub os: serde_json::Value`), so their CONTENTS prove
 * nothing about conformance — while `os.version` is the host kernel string and
 * `runtime.version` is the Node version, both of which differ on every machine.
 * `runtime.name` is absent on purpose: it is an SDK constant, not a host value.
 */
const HOST_DERIVED = new Set([
  'os.name',
  'os.version',
  'runtime.version',
  'device.family',
  'device.model',
  'device.arch',
]);

function normalizeValue(key: string, parentKey: string, value: unknown): unknown {
  if (typeof value === 'string') {
    if (HOST_DERIVED.has(`${parentKey}.${key}`)) return '<host>';
    if (key in FRAME_IDENTITY) return FRAME_IDENTITY[key];
    switch (key) {
      case 'timestamp':
      case 'sent_at':
        return FIXED.timestamp;
      case 'event_id':
        return FIXED.eventId;
      case 'session_id':
        return FIXED.sessionId;
      case 'device_id':
        return FIXED.deviceId;
      case 'workflow_id':
        return FIXED.workflowId;
      case 'raw_stacktrace':
        return '<normalized>';
      case 'build_id':
      case 'isolate_dso_base':
        return '<normalized>';
      default:
        return value;
    }
  }
  if (typeof value === 'number') {
    if (key === 'lineno') return FIXED.lineno;
    if (key === 'colno') return FIXED.colno;
  }
  // `null` is left alone everywhere: nullability is part of what the fixture
  // proves, so a null must never be papered over with a placeholder.
  return value;
}

function normalize(node: unknown, key = '', parentKey = ''): unknown {
  // Array children keep their container's key AND its parent, so a frame inside
  // `stacktrace: [...]` is still seen as living under `stacktrace`.
  if (Array.isArray(node)) return node.map((n) => normalize(n, key, parentKey));
  if (node !== null && typeof node === 'object') {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(node as Record<string, unknown>)) {
      out[k] = normalize(v, k, key);
    }
    return out;
  }
  return normalizeValue(key, parentKey, node);
}

/** Write one captured envelope as this SDK's committed wire fixture. */
export function writeWireFixture(sdk: string, envelope: unknown): void {
  const path = fileURLToPath(new URL(`../../wire-fixtures/${sdk}.json`, import.meta.url));
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, `${JSON.stringify(normalize(envelope), null, 2)}\n`, 'utf8');
}
